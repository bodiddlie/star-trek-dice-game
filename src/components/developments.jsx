import React from 'react';
import { connect } from 'react-redux';

const Developments = props => {
  const { developments } = props;

  return (
    <div>
      {developments.map(d =>
        <span key={d.id}>
          {d.title}
        </span>
      )}
    </div>
  );
};

const mapStateToProps = state => {
  return {
    developments: state.developments.availableDevelopments,
  };
};

export default connect(mapStateToProps)(Developments);
