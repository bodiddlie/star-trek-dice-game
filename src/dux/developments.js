import { cards } from './development-deck';
import { clone, shuffle } from './util';

const SHUFFLE_DECK = '[Developments] Shuffle Deck';
const DRAW_DEVELOPMENT = '[Developments] Draw Development';
const MAKE_DEVELOPMENT_AVAILABLE = '[Devlopments] Make Development Available';
const DISCARD_DEVELOPMENT = '[Developments] Discard Development';
const CLAIM_DEVELOPMENT = '[Developments] Claim Development';
const CLEAR_DRAWN_DEVELOPMENT = '[Developments] Clear Drawn Development';

export function shuffleDeck() {
  return { type: SHUFFLE_DECK };
}

export function drawDevelopment() {
  return { type: DRAW_DEVELOPMENT };
}

export function makeDevelopmentAvailable(development) {
  return { type: MAKE_DEVELOPMENT_AVAILABLE, development };
}

export function discardDevelopment(development) {
  return { type: DISCARD_DEVELOPMENT, development };
}

export function claimDevelopment(development) {
  return { type: CLAIM_DEVELOPMENT, development };
}

export function clearDrawnDevelopment() {
  return { type: CLEAR_DRAWN_DEVELOPMENT };
}

export const initialState = {
  deck: clone(cards),
  drawnDevelopment: null,
  availableDevelopments: [],
  discardedDevelopments: [],
  claimedDevelopments: [],
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case SHUFFLE_DECK: {
      return { ...state, deck: shuffle(state.deck) };
    }
    case DRAW_DEVELOPMENT: {
      const drawnDevelopment = state.deck[0];
      const deck = [...state.deck.slice(1)];
      return { ...state, deck, drawnDevelopment };
    }
    case MAKE_DEVELOPMENT_AVAILABLE: {
      const availableDevelopments = [...state.availableDevelopments, action.development];
      return { ...state, availableDevelopments };
    }
    case DISCARD_DEVELOPMENT: {
      const discardedDevelopments = [...state.discardedDevelopments, action.development];
      return { ...state, discardedDevelopments };
    }
    case CLAIM_DEVELOPMENT: {
      const claimedDevelopments = [...state.claimedDevelopments, action.development];
      return { ...state, claimedDevelopments };
    }
    case CLEAR_DRAWN_DEVELOPMENT: {
      return { ...state, drawnDevelopment: null };
    }
    default: {
      return state;
    }
  }
}
