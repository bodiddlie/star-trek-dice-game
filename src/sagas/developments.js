import { put, select } from 'redux-saga/effects';

import { drawDevelopment, makeDevelopmentAvailable } from '../dux/developments';

export function* startDevelopments() {
  for (let i = 0; i < 3; i++) {
    yield put(drawDevelopment());
    const drawn = yield select(selectors.drawn);
    yield put(makeDevelopmentAvailable(drawn));
  }
}

//----------------------
// SELECTORS
//----------------------

export const selectors = {
  drawn: state => state.developments.drawnDevelopment,
};
