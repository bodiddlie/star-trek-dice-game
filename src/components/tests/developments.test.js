import React from 'react';
import { shallow } from 'enzyme';

import { Developments } from '../developments';

describe('<Developments />', () => {
  it('should render an empty div with no developments', () => {
    const wrapper = shallow(<Developments developments={[]} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a lits of developments', () => {
    const devs = [{ id: 1, title: 'dev 1' }, { id: 2, title: 'dev 2' }];
    const wrapper = shallow(<Developments developments={devs} />);
    expect(wrapper).toMatchSnapshot();
  });
});
