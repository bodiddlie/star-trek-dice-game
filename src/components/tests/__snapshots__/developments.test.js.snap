// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Developments /> should render a lits of developments 1`] = `
<div>
  <span>
    dev 1
  </span>
  <span>
    dev 2
  </span>
</div>
`;

exports[`<Developments /> should render an empty div with no developments 1`] = `<div />`;
